### 基金的季度报告查找下载   
2021年4月19日


In [1]:
import requests
import random
from bs4 import BeautifulSoup as bs
import time
#import redis
import re
import json
from tqdm.notebook import tqdm 
import pandas as pd
from selenium import webdriver
from selenium.webdriver.common.action_chains import ActionChains
import pandas as pd
header = {"User-Agent": 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) '
                                'Chrome/78.0.3904.108 Safari/537.36'}

def downloadFILE_old(url,name):
    try:
        resp = requests.get(url=url,headers=headers,stream=True)
        content_size = int(resp.headers['Content-Length'])/1024
        with open(name, "wb") as f:
            #print("Pkg total size is:",content_size,'k,start...')
            print("文件%s的大小是：%.2fK,开始下载..." % (name,content_size))
            for data in tqdm(iterable=resp.iter_content(1024),total=content_size,unit='k',desc=name):
                f.write(data)
        print(name + "下载完成！")
    except Exception as e:
        print(name + "下载失败！", e)

In [36]:
## 读取基金页面的规模
def get_jjgm(jjcode):
    timeout=10
    url = f'http://fund.eastmoney.com/{jjcode}.html'
    try:
        page = requests.get(url, headers=header, timeout=timeout)
        page.encoding = 'utf-8'
        result = ('success', page.text)
    except (requests.exceptions.ConnectionError, requests.exceptions.Timeout, requests.exceptions.HTTPError):
        result = ('error', url)
    page_context = result[1]
    fund_scale = re.search(r'基金规模</a>：((?:\d+(?:\.\d{2}|)|--)亿元.*?)<', page_context)
    print(fund_scale.group(1))
get_jjgm("168104")    

0.73亿元（2021-03-31）


In [6]:
df_fund = pd.read_excel("300502基金持股.xls")

In [ ]:
fund_scale = re.search(r'基金规模</a>：((?:\d+(?:\.\d{2}|)|--)亿元.*?)<', page_context)

In [7]:
df_fund.code=df_fund.code.map(lambda x: "%06d"%x)
fund_list = df_fund.code.tolist()

In [8]:
#fund_list = df_fund.sort_values(by="s2004",ascending=False).head(40).code.tolist()
fund_list = df_fund.sort_values(by="s2004",ascending=False).code.tolist()

In [3]:
browser = webdriver.Chrome()
jjcode = "009062"
url = "http://fundf10.eastmoney.com/ccbdxq_%s_300502.html" % jjcode

In [4]:
def check_cg(jjcode,stockid):  #查找2020年12月31日到2021年1月31日两次季报对stockid的持股是否有增加
    url = "http://fundf10.eastmoney.com/ccbdxq_%s_300502.html" % jjcode
    browser.get(url)
    title_element = browser.find_elements_by_xpath("//h4[@class='title']")
    if len(title_element)>0:
        title = title_element[0].text
    target = browser.find_elements_by_xpath("//div[@id='cctable']")
    df = pd.read_html(target[0].get_attribute("outerHTML"))[0]
    if df.loc[0,"报告期"]=="2021-03-31" :
        s2004 = df.loc[df["报告期"]=="2020-12-31","持股数（万股）"].values[0]
        s2101 = df.loc[df["报告期"]=="2021-03-31","持股数（万股）"].values[0]
        print(f"基金 {title} 持有{stockid}的变动情况为：从{s2004}万股到今年一季报的{s2101}万股，增加了{round(s2101-s2004,2)}万股  ")
        #print(df)
    else:
        pass
        #print(f"基金{title}的一季报还没出")
check_cg("009970","300502")            

基金 财通内需增长12个月定开混合 (009970) 持有300502的变动情况为：从113.55万股到今年一季报的113.55万股，增加了0.0万股  


In [8]:
for i in fund_list:
    try:
        check_cg(i,"300502")
    except:
        pass

基金 银华同力精选混合 (009394) 持有300502的变动情况为：从262.03万股到今年一季报的419.0万股，增加了156.97万股  
基金 财通内需增长12个月定开混合 (009970) 持有300502的变动情况为：从113.55万股到今年一季报的113.55万股，增加了0.0万股  
基金 银华创业板两年定期开放混合 (161838) 持有300502的变动情况为：从76.98万股到今年一季报的132.88万股，增加了55.9万股  
基金 中邮信息产业灵活配置混合 (001227) 持有300502的变动情况为：从62.18万股到今年一季报的95.18万股，增加了33.0万股  
基金 富国中证500指数(LOF) (161017) 持有300502的变动情况为：从54.85万股到今年一季报的63.15万股，增加了8.3万股  
基金 财通科创主题3年封闭混合 (501085) 持有300502的变动情况为：从155.19万股到今年一季报的155.19万股，增加了0.0万股  
基金 银华成长先锋混合 (180020) 持有300502的变动情况为：从17.01万股到今年一季报的46.62万股，增加了29.61万股  
基金 华夏中证5G通信主题ETF (515050) 持有300502的变动情况为：从420.74万股到今年一季报的426.7万股，增加了5.96万股  
基金 南方信息创新混合A (007490) 持有300502的变动情况为：从769.68万股到今年一季报的592.11万股，增加了-177.57万股  
基金 南方多利增强债券C (202102) 持有300502的变动情况为：从3.79万股到今年一季报的3.79万股，增加了0.0万股  
基金 中邮创新优势灵活配置混合 (001275) 持有300502的变动情况为：从13.5万股到今年一季报的18.5万股，增加了5.0万股  
基金 南方养老2040三年持有混合(F (009572) 持有300502的变动情况为：从1.89万股到今年一季报的1.89万股，增加了0.0万股  
基金 建信中证1000指数增强A (006165) 持有300502的变动情况为：从1.02万股到今年一季报的2.08万股，增加了1.06万股  
基金 富国中证1000指数增强 (161039) 持有

基金 南方信息创新混合A (007490) 持有300502的变动情况为：从769.68万股到今年一季报的592.11万股，增加了-177.57万股  
基金 财通成长优选混合 (001480) 持有300502的变动情况为：从283.88万股到今年一季报的283.88万股，增加了0.0万股  
基金 财通价值动量混合 (720001) 持有300502的变动情况为：从283.88万股到今年一季报的283.88万股，增加了0.0万股  
基金 银华同力精选混合 (009394) 持有300502的变动情况为：从262.03万股到今年一季报的419.0万股，增加了156.97万股  
基金 财通科创主题3年封闭混合 (501085) 持有300502的变动情况为：从155.19万股到今年一季报的155.19万股，增加了0.0万股  
基金 财通内需增长12个月定开混合 (009970) 持有300502的变动情况为：从113.55万股到今年一季报的113.55万股，增加了0.0万股  
基金 银华创业板两年定期开放混合 (161838) 持有300502的变动情况为：从76.98万股到今年一季报的132.88万股，增加了55.9万股  
基金 中邮信息产业灵活配置混合 (001227) 持有300502的变动情况为：从62.18万股到今年一季报的95.18万股，增加了33.0万股  
基金 富国中证500指数(LOF) (161017) 持有300502的变动情况为：从54.85万股到今年一季报的63.15万股，增加了8.3万股  
基金 财通集成电路产业股票A (006502) 持有300502的变动情况为：从39.55万股到今年一季报的39.55万股，增加了0.0万股  
基金 财通多策略福鑫定开混合 (501046) 持有300502的变动情况为：从32.67万股到今年一季报的32.67万股，增加了0.0万股  
基金 银华成长先锋混合 (180020) 持有300502的变动情况为：从17.01万股到今年一季报的46.62万股，增加了29.61万股  
基金 中邮创新优势灵活配置混合 (001275) 持有300502的变动情况为：从13.5万股到今年一季报的18.5万股，增加了5.0万股  
基金 富国中证1000指数增强 (161039) 持有300502的变动情况为：从4.25万股到今年一季报的6.07万股，增加了1.82万股  
基金 南方多利增强债券C (202102) 持有300502的变动情况为：从3.79万股到今年一季报的3.79万股，增加了0.0万股  
基金 南方养老2040三年持有混合(F (009572) 持有300502的变动情况为：从1.89万股到今年一季报的1.89万股，增加了0.0万股  
基金 建信中证1000指数增强A (006165) 持有300502的变动情况为：从1.02万股到今年一季报的2.08万股，增加了1.06万股  
基金 红土创新沪深300增强A (006698) 持有300502的变动情况为：从0.07万股到今年一季报的0.43万股，增加了0.36万股  

In [146]:
#target = browser.find_elements_by_xpath("//div[@id='cctable']")
title = browser.find_elements_by_xpath("//h4[@class='title']")

In [149]:
title[0].text

'银华创业板两年定期开放混合 (161838)'

In [88]:
df = pd.read_html(target[0].get_attribute("outerHTML"))


In [108]:
temp = df[0].loc[0,"报告期"]

In [111]:
print(f"基金{temp}的一季报还没出")

基金f2020-12-31的一季报还没出


In [103]:
len(df[0][df[0]["报告期"]=="2020-12-31"])

1

In [66]:
def check_fund(fundcode):  # 看看基金是否有一季报
    url = "http://fundf10.eastmoney.com/jjgg_%s_3.html" % fundcode
    browser.get(url)
    #soup = bs(browser.page_source, "html.parser")
    target = browser.find_elements_by_xpath("//td/a[contains(text(),'2021年第1季度报告')]")  #找出含有季度报告的数据 表格里的A
    if len(target)>0:
        print(fundcode)

In [67]:
for i in fund_list:
    check_fund(i)

WebDriverException: Message: chrome not reachable
  (Session info: chrome=89.0.4389.128)


In [43]:


check_fund("005827")

005827


In [9]:
# 获取所有股票的名字，存入当地文件
 
def get_stock_names():
    """
    通过东方财富网上爬取股票的名称代码,并存入redis数据库和本地txt文档
    """
    #rds = redis.from_url('redis://:666666@192.168.3.98:6379', db=1, decode_responses=True)   # 连接redis db1
 
    url = "http://quote.eastmoney.com/stocklist.html"
    headers = {
            'Referer': 'http://quote.eastmoney.com/center/gridlist.html',
            'User-Agent': 'Mozilla/5.0 (Windows NT 6.1; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/75.0.3770.100 Safari/537.36'
        }
 
    response = requests.get(url, headers=headers).content.decode('utf-8')   # 网站编码为gbk 需要解码
    #print(response)
    soup = bs(response, 'lxml')
    #print(soup.text)
    all_ul = soup.find('table', id='table_wrapper-table').find_all('ul')   # 获取两个ul 标签数据
    with open('stock_names.txt', 'w+', encoding='utf-8') as f: 
        for ul in all_ul:
            all_a = ul.find_all('a')            # 获取ul 下的所有的a 标签
            for a in all_a:
                #rds.rpush('stock_names', a.text)       # a.text 为a标签中的text数据  rpush将数据右侧插入数据库
                print(a.text)
                f.write(a.text + '\n')
get_stock_names()                

In [7]:
url = "http://fundf10.eastmoney.com/jjgg_007490_3.html"
headers = {
        'Referer': 'http://quote.eastmoney.com/center/gridlist.html',
        'User-Agent': 'Mozilla/5.0 (Windows NT 6.1; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/75.0.3770.100 Safari/537.36'
    }

response = requests.get(url, headers=headers).content.decode('utf-8')   # 网站编码为gbk 需要解码
#print(response)
soup = bs(response, 'lxml')

In [10]:
with open('temp', 'w', encoding='utf-8') as f: 
    f.write(response)

In [11]:
driver = webdriver.Ie()
driver.get(url)

WebDriverException: Message: 'IEDriverServer.exe' executable needs to be in PATH. Please download from http://selenium-release.storage.googleapis.com/index.html and read up at https://github.com/SeleniumHQ/selenium/wiki/InternetExplorerDriver


In [35]:
#chrome_path = "E:\\Download\\ChromeDownload\\chromedriver_win32\\chromedriver.exe"
browser = webdriver.Chrome()

In [37]:
url = "http://fundf10.eastmoney.com/jjgg_%s_3.html" % "007490"
browser.get(url)
target = browser.find_elements_by_xpath("//td/a[contains(text(),'2021年1季度报告')]") 
if len(target)>0:
    print("007490")

In [34]:
 #driver.find_elements_by_xpath("//*[contains(text(), 'My Button')]")
 #target = browser.find_elements_by_xpath("//td/a[contains(text(),'季度报告')]/following-sibling::a")
browser.get(url)

NoSuchWindowException: Message: Unable to get browser


In [72]:
target[0].find_element_by_xpath("./following-sibling::a").get_attribute("href")  #找出下一个相邻的A

'http://pdf.dfcfw.com/pdf/H2_AN202101221453171343_1.pdf'

In [83]:
downloadFILE_old("http://pdf.dfcfw.com/pdf/H2_AN202101221453171343_1.pdf","1.pdf")

文件1.pdf的大小是：518.55K,开始下载...



1.pdf下载完成！


In [61]:
# 2021年4月20日
# 在使用fundcrawler.py生成相应的文件后，读取相应的基金文件进行排序
def p(x):
    if "-" in x:
        return 0
    else:
        return float(x.strip("%"))
df = pd.read_csv("FundCrawler/results/混合型.csv",engine="python")
for i in ["近1月","近3月","近6月","近1年","近3年","成立来"]:
    df[i]=df[i].map(p)

In [62]:
df["bj"]=df["近3月"]+df["近6月"]+df["近1年"]+df["近3年"]

In [63]:
df.sort_values(by="bj",ascending=False)

基金名称    基金代码                  基金规模    近1月    近3月    近6月  \
843     前海开源中国稀缺资产混合A    1679   30.35亿元（2020-12-31）   2.98   0.00  20.99   
933     前海开源中国稀缺资产混合C    2079   11.94亿元（2020-12-31）   2.97   0.00  20.96   
633         易方达新收益混合A    1216   37.58亿元（2021-03-31）   8.92   0.00  22.75   
222          农银研究精选混合     336   27.43亿元（2020-12-31）   5.59   0.00  21.63   
2071         交银股息优化混合    4868    7.39亿元（2020-12-31）   6.49   3.07  25.38   
793         易方达新收益混合C    1217   18.64亿元（2021-03-31）   8.87   0.00  22.64   
135         前海联合泓鑫混合A    2780    6.35亿元（2020-12-31）   8.66   0.00  18.76   
308           长安裕泰混合C    5342    0.77亿元（2020-12-31）   5.85   0.00  25.40   
1641        中欧价值智选混合E    1887    0.53亿元（2020-12-31）   1.29  13.37  29.49   
1654        中欧价值智选混合A  166019    4.52亿元（2020-12-31）   1.30  13.37  29.50   
903       易方达瑞恒灵活配置混合    1832   23.88亿元（2021-03-31）   4.10   0.00  24.27   
95          农银工业4.0混合    1606   39.35亿元（2020-12-31）   6.50   0.00  22.96   
1686        中欧价值智选混合C    4235    0.42亿元（2020-12-31）   1.23  13.16  29.00   
740      易方达瑞程灵活配置混合C    3962    4.96亿元（2021-03-31）   2.53   0.00  15.88   
19            农银新能源主题    2190  153.14亿元（2020-12-31）  10.60   0.00  24.82   
578      易方达瑞程灵活配置混合A    3961   13.89亿元（2021-03-31）   2.55   0.00  16.00   
264         易方达环保主题混合    1856   30.54亿元（2021-03-31）   4.52   0.00  15.70   
1352        中欧时代智慧混合A    5241   13.65亿元（2020-12-31）   2.22   0.00  31.02   
1645         华安文体健康混合    1532    1.96亿元（2020-12-31）   5.32   8.98  27.63   
968       大成睿景灵活配置混合A    1300   10.54亿元（2020-12-31）   0.06   8.83  30.66   
311         长安鑫旺价值混合C    5050    1.11亿元（2020-12-31）   6.83   0.00  24.84   
435         长安鑫旺价值混合A    5049    0.52亿元（2020-12-31）   6.84   0.00  24.94   
893            泓德泓华混合    2846    8.18亿元（2020-12-31）   5.84   0.00  17.99   
1497          汇安丰泽混合A    3889    2.21亿元（2020-12-31）   4.35   2.84  21.77   
619          新华优选消费混合  519150    1.89亿元（2020-12-31）  14.57   2.14  29.40   
472           南华丰淳混合A    5296    0.03亿元（2020-12-31）   5.80  24.25  46.87   
797     前海开源国家比较优势混合A    1102   49.04亿元（2020-12-31）   2.97   0.00  21.63   
85           金鹰改革红利混合    1951    3.88亿元（2020-12-31）   2.77   3.44  45.83   
348          诺德周期策略混合  570008   11.89亿元（2020-12-31）   9.78   0.00  23.55   
1137    景顺长城鼎益混合(LOF)  162605  136.23亿元（2020-12-31）   7.38   0.00  21.36   
...               ...     ...                   ...    ...    ...    ...   
2524           永赢鑫欣混合   10923    2.51亿元（2021-01-14）   0.73   0.00   0.00   
2350    中融景颐6个月持有期混合A   10683   10.32亿元（2021-01-27）   0.26   0.00   0.00   
2347        鹏华汇智优选混合A   10894  113.03亿元（2021-01-08）   1.57   0.00   0.00   
2346    国金鑫新灵活配置(LOF)  501000    0.01亿元（2020-12-31）   0.00   0.00   0.00   
630         博时产业精选混合C   10456    2.24亿元（2020-11-06）   0.20   0.00   0.00   
2274        鹏华成长智选混合A   10264  124.90亿元（2020-12-31）   2.06   0.00   0.00   
648      东方红启瑞三年持有混合A  910011              --亿元（--）   0.00   0.00   0.00   
641   财通资管优选回报一年持有期混合    9774   18.31亿元（2020-12-31）   3.59   0.00   0.00   
639     广发创新医疗两年持有混合A   10731    5.04亿元（2021-03-19）   4.69   0.00   0.00   
2294     交银内需增长一年持有混合   10454   49.50亿元（2020-12-11）   4.66   0.00   0.00   
2297       国寿安保高股息混合C    9501    0.01亿元（2020-12-31）   0.00   0.00   0.00   
2299        鹏华汇智优选混合C   10895    5.84亿元（2021-01-08）   1.51   0.00   0.00   
2301        兴银策略智选混合C   10428    1.81亿元（2020-11-10）   0.07   0.00   0.00   
3197       银河臻选多策略混合C    8387    0.03亿元（2020-12-31）   0.09   0.00   0.00   
2302        工银灵动价值混合C   10745    2.88亿元（2020-12-25）   1.03   0.00   0.00   
3195     农银策略收益一年持有混合   10347   78.85亿元（2021-01-27）   0.00   0.00   0.00   
2306    创金合信景雯灵活配置混合A   10597    0.86亿元（2021-02-19）   1.49   0.00   0.00   
3179        大成丰享回报混合A    9653    3.12亿元（2020-11-18）   0.89   0.00   0.00   
117       创金合信竞争优势混合A   11206   25.03亿元（2021-01-25）   2.29   0.00   0.00   
628        鹏华高质量增长混合A   10490   31.22亿元（2020-11-18）   3.08   0.00   0.00   
626    富国融泰三个月定开混合发起式   10400   10.10亿元（2021-02-03）   0.00   0.00